### ISSUE TITLE
torch.pow type promotion is incorrect with 0-dimensional tensors

### ISSUE THREAD

**Issue by gchanan** [pytorch/pytorch#50121]
## 🐛 Bug

The issue is that it treats python numbers and 0-dim scalars as having the same priority class, when the 0-dim scalar should be favored.  This is a little difficult to see when not using the `out=` variant (because it calculates the result type correctly), but it's obvious with out variants:
```
>>> torch.pow(2, torch.tensor([17], dtype=torch.uint8), out=torch.tensor([0]))
tensor([0])
>>> torch.pow(2, torch.tensor(17, dtype=torch.uint8), out=torch.tensor(0))
tensor(131072)
```

Note the top example is "correct" in that it is doing the computation in uint8 and then casting the output to `torch.int64`

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.8 (master)
 - OS (e.g., Linux): Linux

## Additional context

cc @ezyang @gchanan @mruberry @rgommers @heitorschueroff @nairbv

**Comment by gchanan** [comment]
Interestingly, this doesn't apply to CUDA:
```
>>> torch.pow(2, torch.tensor(17, dtype=torch.uint8, device='cuda'), out=torch.tensor(0, device='cuda'))
tensor(131072, device='cuda:0')
>>> torch.pow(2, torch.tensor([17], dtype=torch.uint8, device='cuda'), out=torch.tensor([0], device='cuda'))
tensor([131072], device='cuda:0')
```
